### ISSUE TITLE
Get counting loss working on a toy example

### ISSUE THREAD

**Issue by githubharald** [SchroeterJulien/AAAI-2021-Learning-Precise-Temporal-Point-Event-Detection-with-Misaligned-Labels#1]
Hi Julien,

we're working on event detection in time-series data for sports.
Looking for more suitable loss functions than just computing cross-entropy per time-step we came across your papers about the smoothing and counting losses.

To check if the losses do their job we tried to train a simple model on a toy-problem: take a 1D sequence with a triangular function and train the model to predict its peaks.
We randomly move the target labels a bit in time while training to account for noisy training labels.

When using only the smoothing loss we get what we expect - a prediction usually spread over 3 time-steps centered around the correct time-step.
It is much more peaked compared to smoothing only the labels (and not the predictions), but still it is spread over 3 time-steps.

This is why we wanted to additionally add the counting loss to get really sharp peaks (where ideally only 1 time-step encodes the event).
However, we did not succed with this.

We implemented the counting loss using the recursive formula (code see below). 
And we compute the total loss by adding the smoothing loss and the counting loss together, with the counting loss weight factor increasing from 0 to 0.01 during training.
Still, training quickly get's stuck at some loss value, and it does not decrease from then on.
Also visually it can be seen that the events are still spread over 3 time-steps.

* Did we miss something?
* Are there some critical parts we have to take care of, e.g. the weighting scheme of the two losses?
* What should be the output activation function? The counting loss expects values between 0 and 1, suggesting a sigmoid activation, but the squared error loss should be used with raw outputs instead
* Even when using the recursive formula, the counting loss is quite slow (for length 10 sequences!). Any other way to speed things up? 

Code for counting loss, where ps is the sequence of predictions and ks the counts (list of ints) and k (int) the count for the current batch element:
````python
def compute_counting_loss_recursive(ps, k, t):
    if t == -1:
        return 1 if k == 0 else 0

    if k == 0:
        return (1 - ps[..., t]) * compute_counting_loss_recursive(ps, k, t - 1)

    else:
        return (1 - ps[..., t]) * compute_counting_loss_recursive(ps, k, t - 1) + \
               ps[..., t] * compute_counting_loss_recursive(ps, k - 1, t - 1)


def compute_counting_loss(ps, ks):
    num_timesteps = ps.shape[2]
    batch_size = ps.shape[0]

    loss = 0
    ps = torch.sigmoid(ps)
    eps = np.finfo(float).eps
    for b in range(batch_size):
        curr_ps = ps[b:b + 1]
        curr_k = ks[b]
        curr_loss = compute_counting_loss_recursive(curr_ps, curr_k, num_timesteps - 1)
        curr_loss = -torch.log(torch.sum(curr_loss) + eps)
        loss += curr_loss

    loss = loss / batch_size
    return loss
````

Example:
![P6](https://user-images.githubusercontent.com/15148095/145061821-907d66be-99b1-41d5-947b-78b4a2bef6d1.png)

Best
Harald

_(Of course that's not an issue of your code but it might be interesting for others too so I opened an issue here at github. If your prefer the discussion via email please let me know)._

**Comment by githubharald** [comment]
Some minutes after opening this issue things got much better :-) ... I'm now able to train a model that outputs sharp spikes.
(it was a couple of small issues in the code and I'm using cross entropy for the smoothing loss now).

![image](https://user-images.githubusercontent.com/15148095/145073923-c777f529-a09c-4acd-ad6d-3ca4b3961113.png)


Therefore only one of my questions remains: how to speed-up the loss calculation (as written above, currently using the recursive formula from your 2019 paper)?

Best
Harald

**Comment by SchroeterJulien** [comment]
Hi Harald, 

Sport event detection is an exciting task, and it is great to hear that you might try the counting loss for it.

First, using the cross-entropy or the MSE for the smooth loss shouldn't have a huge impact on performance (if the smoothing parameter is chosen big enough). If the smoothing parameter is set too low, then you might run into trouble with the cross-entropy since the log-nature of the loss could penalize too much predictions that are far off from the label (which might happen when it is really noisy). Overall, we chose the MSE simply to be consistent with the literature on human pose estimation and facial landmark detection. 

Regarding **efficiency**, the counting loss will never be as fast as e.g., a cross-entropy. Neverthless, I used the loss on application with over 1000 times steps (see image application  [here](https://github.com/SchroeterJulien/ACCV-2020-Subpixel-Point-Localization)) and it was still running decently on a TitanXp. Overall, while your recursive implementation is really elegant (especially when you look at how I implemented it), its _current_ form computes the counting loss with **exponential complexity** which is far from optimal (see figure below). 

![Current_Recursion](https://user-images.githubusercontent.com/50208711/145292022-a18c78a3-2919-43ec-813d-e6ed4ccbe7e6.jpg)

The solution to make it faster is to start from t=0 with a count distribution where all the mass is on k=0, i.e., [1,0,0,0,0]. You can then process one timestep after the other using the recursion formula (see image below). This is significantly faster with a **complexity of O(T*K)**. This method also allows you to process the entire batch at the same time (which might make it even quicker).  (There is a really rough implementation [here](https://github.com/SchroeterJulien/ACCV-2020-Subpixel-Point-Localization/blob/115329d69ccad6d3ade94ac8f17642f66c251069/4_1_Single%20Molecule%20Localization%20Microscopy/pytorch_SmoothedLosses.py#L131) for comparison). 

![Efficient_Recursion](https://user-images.githubusercontent.com/50208711/145292656-68a850a1-e2cd-401e-996e-d4c9f51dc58a.jpg)

There is this additional tricks that I used sometimes to make it faster: (danger zone) 
- Discard the timestep that do not significantly affect the count distribution. You will have to choose your own threshold for this, but it is safe to say that a prediction of 1e-7 won't have any significant impact on the count distribution. Therefore, you can remove them for the loss computation. 

Hope this helps you to make it faster :) 

Do not hesitate to ask any question you have. It is great to see people try out the loss. 

Best, Julien

**Comment by githubharald** [comment]
Hi Julien,

thanks for your great explanation :-). I completely overlooked that my function computed the same values multiple times which of course made it slow.

Implemented it now with the recursive formula "flipped upside down" as you suggested ... it's much faster :-). 

Here's the adapted code:

````python
def fast_counting_loss(ps: torch.Tensor, ks: List[int]) -> torch.Tensor:
    """
    Compute counting loss.

    Args:
        ps: Tensor containing the model output (logits) for the sequence.
        ks: List of ints, containing the target count per batch element.

    Returns:
        The reduced loss value for the batch.
    """
    # apply sigmoid to get probabilities for each time-step
    ps = torch.sigmoid(ps)

    # create k-t grid to compute Pr(Y(t)=k), we need one additional element for both k and t (t is 1-indexed)
    max_b, _, max_t = ps.shape
    max_k = max(ks)
    kt_grid = torch.zeros([max_b, max_k + 1, max_t + 1])
    kt_grid[..., 0, 0] = 1  # probability for a count of 0 in an empty sequence is 1

    # apply recursive formula, but "upside down", starting from k=t=0
    for t in range(1, max_t + 1):
        p = ps[..., t - 1]  # sequence probability predicted by model
        kt_grid[..., t] += (1 - p) * kt_grid[..., t - 1].clone()  # current time-step is a "0"
        kt_grid[..., 1:, t] += p * kt_grid[..., :-1, t - 1].clone()  # current time-step is a "1"

    # select computed Pr(Y(tmax)=k) value for each batch element, and then reduce to a scalar value
    loss = kt_grid[torch.arange(max_b), ks, -1]
    eps = torch.finfo(loss.dtype).eps
    loss = torch.sum(-torch.log(loss + eps)) / max_b
    return loss
````


Best, 
Harald
